src/bootstrap/tool.rs

```diff
@@ -367,6 +367,7 @@ bootstrap_tool!(
     RustdocTheme, "src/tools/rustdoc-themes", "rustdoc-themes";
     ExpandYamlAnchors, "src/tools/expand-yaml-anchors", "expand-yaml-anchors";
     LintDocs, "src/tools/lint-docs", "lint-docs";
+    JsonDocCk, "src/tools/jsondocck", "jsondocck", is_unstable_tool = true;
```

Comment:
`is_unstable` is used to tell if `RUSTC_BOOTSTRAP` should be passed. I don't think there's any reason to use RUSTC_BOOTSTRAP here.
```suggestion
    JsonDocCk, "src/tools/jsondocck", "jsondocck";
```
(I'll see about removing this from compile test soon, then we could get rid of the option altogether)

Reply:
Hmm, this was adding in 51b7044347d2cb9e2eae18ad49c79803dda60be2, which IIRC was related to something with bootstrapping libtest: https://github.com/rust-lang/rust/pull/68019. And it turns out that is actually necessary because the policy is to build against master libstd when using RUSTC_BOOTSTRAP for anything other than the compiler: https://github.com/rust-lang/rust/issues/59264

@cuviper this confused me at first because building with `Mode::ToolBootstrap` actually still allows using unstable features. Would it be possible to add `RUSTC_BOOTSTRAP` only when building in mode `ToolStd`?

Reply:
See https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Removing.20unnecessary.20RUSTC_BOOTSTRAP, this ended up harder to change than I expected.

Reply:
In principle, yes I agree that those would preferably not be allowed to use unstable features at all. In practice, that may be hard to isolate... as you found.

Reply:
Trying to remove `is_unstable_tool` triggers an assertion failure:
`thread 'main' panicked at 'assertion failed: !use_snapshot || stage == 0 || self.local_rebuild', src\bootstrap\builder.rs:1003:9`

Reply:
That's super weird, ToolBootstrap tools should always be compiled with stage 0. Can you post the backtrace?

Reply:
<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'main' panicked at 'assertion failed: !use_snapshot || stage == 0 || self
.local_rebuild', src\bootstrap\builder.rs:1008:9
stack backtrace:
   0:     0x7ff61c3b9285 - std::backtrace_rs::backtrace::dbghelp::trace
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\..\..\backtrace\src\backtrace\dbghelp.rs:98
   1:     0x7ff61c3b9285 - std::backtrace_rs::backtrace::trace_unsynchronized
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\..\..\backtrace\src\backtrace\mod.rs:66
   2:     0x7ff61c3b9285 - std::sys_common::backtrace::_print_fmt
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\sys_common\backtrace.rs:67
   3:     0x7ff61c3b9285 - std::sys_common::backtrace::_print::{{impl}}::fmt
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\sys_common\backtrace.rs:46
   4:     0x7ff61c3d9e9b - core::fmt::write
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\core\src\fmt\mod.rs:1078
   5:     0x7ff61c3b2afd - std::io::Write::write_fmt<std::sys::windows::stdio::S
tderr>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\io\mod.rs:1519
   6:     0x7ff61c3bcf6d - std::sys_common::backtrace::_print
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\sys_common\backtrace.rs:49
   7:     0x7ff61c3bcf6d - std::sys_common::backtrace::print
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\sys_common\backtrace.rs:36
   8:     0x7ff61c3bcf6d - std::panicking::default_hook::{{closure}}
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:208
   9:     0x7ff61c3bca7a - std::panicking::default_hook
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:225
  10:     0x7ff61c3bd8ae - std::panicking::rust_panic_with_hook
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:591
  11:     0x7ff61c3bd3a3 - std::panicking::begin_panic_handler::{{closure}}
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:495
  12:     0x7ff61c3b9c5f - std::sys_common::backtrace::__rust_end_short_backtrac
e<closure-0,!>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\sys_common\backtrace.rs:141
  13:     0x7ff61c3bd329 - std::panicking::begin_panic_handler
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:493
  14:     0x7ff61c3d86e0 - core::panicking::panic_fmt
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\core\src\panicking.rs:92
  15:     0x7ff61c3d862c - core::panicking::panic
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\core\src\panicking.rs:50
  16:     0x7ff61baeabb9 - bootstrap::builder::Builder::cargo
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:1008
  17:     0x7ff61befb922 - bootstrap::tool::prepare_tool_cargo
                               at D:\Dev\Git\rust\src\bootstrap\tool.rs:230
  18:     0x7ff61bef9cf9 - bootstrap::tool::{{impl}}::run
                               at D:\Dev\Git\rust\src\bootstrap\tool.rs:61
  19:     0x7ff61bddda62 - bootstrap::builder::Builder::ensure<bootstrap::tool::
ToolBuild>
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:1494
  20:     0x7ff61bf01751 - bootstrap::tool::{{impl}}::run
                               at D:\Dev\Git\rust\src\bootstrap\tool.rs:326
  21:     0x7ff61be4b1c7 - bootstrap::builder::Builder::ensure<bootstrap::tool::
JsonDocCk>
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:1494
  22:     0x7ff61bee4c1e - bootstrap::test::{{impl}}::run
                               at D:\Dev\Git\rust\src\bootstrap\test.rs:1075
  23:     0x7ff61bdbedb1 - bootstrap::builder::Builder::ensure<bootstrap::test::
Compiletest>
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:1494
  24:  24:     0x7ff61bef3ad1 - bootstrap::test::{{impl}}::run
                               at D:\Dev\Git\rust\src\bootstrap\test.rs:852
  25:     0x7ff61be40575 - bootstrap::builder::Builder::ensure<bootstrap::test::
RustdocJson>
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:1494
  26:     0x7ff61bef3957 - bootstrap::test::{{impl}}::make_run
                               at D:\Dev\Git\rust\src\bootstrap\test.rs:848
  27:     0x7ff61bae30cc - bootstrap::builder::StepDescription::maybe_run
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:179
  28:     0x7ff61bae38df - bootstrap::builder::StepDescription::run
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:217
  29:     0x7ff61bae7a1f - bootstrap::builder::Builder::run_step_descriptions
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:570
  30:     0x7ff61bae78a3 - bootstrap::builder::Builder::execute_cli
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:561
  31:     0x7ff61b904334 - bootstrap::Build::build
                               at D:\Dev\Git\rust\src\bootstrap\lib.rs:506
  32:     0x7ff61b8f2083 - bootstrap::main
                               at D:\Dev\Git\rust\src\bootstrap\bin\main.rs:30
  33:     0x7ff61b8f1c0b - core::ops::function::FnOnce::call_once<fn(),tuple<>>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\library\core\src\ops\function.rs:227
  34:     0x7ff61b8f2acb - std::sys_common::backtrace::__rust_begin_short_backtr
ace<fn(),tuple<>>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\library\std\src\sys_common\backtrace.rs:125
  35:     0x7ff61b8f1ab1 - std::rt::lang_start::{{closure}}<tuple<>>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\library\std\src\rt.rs:66
  36:     0x7ff61c3bdba4 - core::ops::function::impls::{{impl}}::call_once
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\library\core\src\ops\function.rs:280
  37:     0x7ff61c3bdba4 - std::panicking::try::do_call
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:379
  38:     0x7ff61c3bdba4 - std::panicking::try
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:343
  39:     0x7ff61c3bdba4 - std::panic::catch_unwind
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panic.rs:396
  40:     0x7ff61c3bdba4 - std::rt::lang_start_internal
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\rt.rs:51
  41:     0x7ff61b8f1a83 - std::rt::lang_start<tuple<>>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\library\std\src\rt.rs:65
  42:     0x7ff61b8f26a0 - main
  43:     0x7ff61c3e1564 - invoke_main
                               at d:\A01\_work\6\s\src\vctools\crt\vcstartup\src
\startup\exe_common.inl:78
  44:     0x7ff61c3e1564 - __scrt_common_main_seh
                               at d:\A01\_work\6\s\src\vctools\crt\vcstartup\src
\startup\exe_common.inl:288
  45:     0x7ffbb4fb7034 - BaseThreadInitThunk
  46:     0x7ffbb6edd0d1 - RtlUserThreadStart
failed to run: D:\Dev\Git\rust\build\bootstrap\debug\bootstrap test src/test/rus
tdoc-json
Build completed unsuccessfully in 0:00:38

Process finished with exit code 1
```

</p>
</details>

Reply:
It looks like it fails because it's not a rebuild, it's during stage 1, and if *not* unstable, it *is* ToolBootstrap, which is apparently not allowed in conjunction with the other two? My first guess is that it relates to compiletest being unstable, but that's not based on any evidence.